components/script/script_thread.rs

```diff
@@ -325,11 +325,51 @@ impl OpaqueSender<CommonScriptMsg> for Sender<MainThreadScriptMsg> {
     }
 }
 
+/// The state that a document managed by this script thread could be in.
+#[must_root]
+enum DocumentState {
+    Kept(bool, JS<Document>),
+    Discarded(ServoUrl),
+}
+
+impl JSTraceable for DocumentState {
+    fn trace(&self, trc: *mut JSTracer) {
+        // Note: we only trace thawed documents.
+        if let DocumentState::Kept(true, ref doc) = *self {
+            doc.trace(trc);
```

Comment:
The derived version would trace documents even if they were `Kept(false, doc)`. The idea is to implement poor-dev's weak references, that don't keep the documents alive.